### ISSUE TITLE
[enhancement]: Add support for Workload Identity Federation when joining agent to a pool

### ISSUE THREAD

**Issue by HarryGwinnell** [microsoft/azure-pipelines-agent#4802]
We're currently using PATs to join the agent to our pool, but in an effort to remove PATs and secrets, we're trying to switch to secretless auth where possible. We currently have our ADO organization connected to Tenant A, but the agents are created in an Azure Subscription managed by Tenant B.

Managed Identities do not work cross-tenant, so federation becomes necessary. The agent does support Service Principal auth, but it requires a Client Secret. Instead, it'd be useful if the Service Principal option allowed a Federated Identity to be used instead. Alternatively, we can produce a federated OAuth token ourselves, if the agent allows passing an OAuth token.

**Comment by DmitriiBobreshev** [comment]
Hi @HarryGwinnell, thanks for the feedback, even if the agent has [Oauth](https://github.com/microsoft/azure-pipelines-agent/blob/master/src/Microsoft.VisualStudio.Services.Agent/Constants.cs#L269) connection, it could not be configured through agent's configuration. 
We are working on more prioritized issues at the moment, but will try to implement Federated Identity soon.

**Comment by HarryGwinnell** [comment]
@DmitriiBobreshev Is there a rough timeline for this? It's blocking some S360 work and may become a hard blocker in some upcoming waves

**Comment by sbebrys** [comment]
@HarryGwinnell we are planning do the same and after some tries I could confirm that access_token gained from manually exchanging client-assertion-token can be used instead PAT (as the same parameter with auth type PAT) and worked well. 

Because our solution is based on Keda scaler, where each agent is running once as job in kubernetes, I also requested Keda to support workload identity there, to give scaler password less access to the ado job queue. Keda implemented that and now we could run agents from different tenant or even cloud.

Still we have challenge how to not expose injected client-assertion-token to the jobs and unregister agent after 1h (final ado access_token expiration time cannot be managed).

It will be great to have this support out of the box, but timeline is unfortunately unknown.

**Comment by tgothorp** [comment]
Hello 👋 Is there any update on this issue? We're also transitioning to federated identity and would greatly benefit from having this available for our build agents.

**Comment by Frulfump** [comment]
Still relevant, the onus on keeping it active is on the agent team imo.

**Comment by cveld** [comment]
I believe az cli caches tokens. So if the agent would allow for customizing the fabrication of the Authorization header we could do the following:

For Entra access token based acess:

```bash
TOKEN=$(az account get-access-token \
  --resource 499b84ac-1321-427f-aa17-267ca6975798 \
  --query accessToken \
  -o tsv)

curl -s \
  -H "Authorization: Bearer $TOKEN" \
  -H "Content-Type: application/json" \
  "https://dev.azure.com/{YOUR_ORG}/_apis/projects?api-version=7.0"
```

And for classic PAT:

```bash
B64=$(echo -n ":abcdef123456..." | base64)

curl -s \
  -H "Authorization: Basic $B64" \
  -H "Content-Type: application/json" \
  "https://dev.azure.com/{YOUR_ORG}/_apis/projects?api-version=7.0"
```

**Comment by OzoneNZ** [comment]
Seems a little odd this hasn't had more MSFT traction given the well-publicized AzDO project to kill off PAT reliance across the rest of the stack?

PAT reliance for agent pool binding seems like the sore thumb/Achilles heel given agent pools are a high-risk entity, in particular the common scenario where PATs expire and tank the entire agent pool and associated build jobs with them.

We've worked around this in-house ourselves via Entra ID bearer tokens and associated hackery as others have covered above, but it really feels like a big miss having PATs this deep amongst AzDO's most critical areas while almost all others now have sanctioned alternatives :(

Beyond achieving native token-based agent binding it would also be great to see a Kubernetes-native supplement that plays nicely with Azure Workload Identity - though we currently home-roll this for now and achieve decent results, it's not quite as reliable nor neat as we'd like.

**Comment by Frulfump** [comment]
Still relevant as always waiting for the maintainers to interact at all.

**Comment by cveld** [comment]
Microsoft has added multi tenant support for federated identity credentials. I guess the ask from OP has been satisfied and this ticket can be closed? See
https://github.com/arsenvlad/entra-cross-tenant-app-fic-managed-identity

**Comment by sbebrys** [comment]
@cveld Not really, yes token itself can be created and federated cross tenant outside agent, but this complicates startup script for agent executable and process for unattaching agent from pool (especial where init token allready expired). Having that bundled into agent itself as additional auth method will simplify solution.

**Comment by cveld** [comment]
I guess the ticket can be closed? We have implemented multi tenant identity federation successfully. Although the keda engine in container apps still requires PAT.
